### PR TITLE
Invite

### PR DESCRIPTION
**Pull request by slvk14**
**STATUS**
READY/IN DEVELOPMENT

**DESCRIPTION**

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [x] code coverage >90%
- [ ] DB indexes added if needed
- [ ] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs
- [x] code checked with Rubocop and offenses are fixed

image

### CODE REVIEW

File: app/controllers/account/users_controller.rb [hunk 1]
```diff
@@ -1,5 +1,6 @@
 class Account::UsersController < ApplicationController
   layout 'dashboard'
+  before_action :is_admin?
```
Comment (igorkasyanchuk) [inline app/controllers/account/users_controller.rb]: bad name of method

require_admin must BE
PLUS do we plan to use this method somewhere else? If yes - method must be in parent class and before_action just needs to be here

File: app/controllers/account/users_controller.rb [hunk 2]
```diff
@@ -56,4 +57,8 @@ def user_params
     params.require(:user).permit(:first_name, :last_name, :email, :password,
                                  :password_confirmation, :role, :organization_id, :avatar)
   end
+
+  def is_admin?
```
Comment (igorkasyanchuk) [inline app/controllers/account/users_controller.rb]: if method with "?" it must retrun true/false this is why it's bad name

File: app/controllers/users/registrations_controller.rb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 class Users::RegistrationsController < Devise::RegistrationsController
-  before_action :configure_permitted_parameters, only: :create
+  # before_action :configure_permitted_parameters, only: :create
```
Comment (igorkasyanchuk) [inline app/controllers/users/registrations_controller.rb]: do you need this comment?

File: app/controllers/users/registrations_controller.rb [hunk 4]
```diff
@@ -13,6 +13,11 @@ def new
   # POST /resource
   def create
     super
+    if @user.own_organization.present?
+      @user.role = 'admin'
+      @user.save
```
Comment (igorkasyanchuk) [inline app/controllers/users/registrations_controller.rb]: put in model

File: app/views/account/users/index.html.haml [hunk 5]
```diff
@@ -1,2 +1,36 @@
...
+          %td 
+            = user.email
+          %td
+            - if user.role == 'member' 
+              = link_to "Delete", account_user_path(user), method: :delete, data: {confirm: "Are you sure?"}
+          %td{ style: "text-align: center" }
+            - if user.invitation_accepted?
+              %p
+                accepted &#10004;
+            - elsif user.role == 'admin'
```
Comment (igorkasyanchuk) [inline app/views/account/users/index.html.haml]: why this elsif is blank?

Comment (igorkasyanchuk) [reply app/views/account/users/index.html.haml]: move logic to helper or model about "status"

File: app/views/account/users/index.html.haml [hunk 6]
```diff
@@ -1,2 +1,36 @@
...
+        %tr
+          %th{ style: "width:40%;" } Name
+          %th E-mail
+          %th Options
+          %th{ style: "text-align: center" } Status
+      %tbody
+      - @users.each do |user|
+        %tr
+          %td 
+            - if user.role == 'admin'
```
Comment (igorkasyanchuk) [inline app/views/account/users/index.html.haml]: if user.admin?

File: app/views/account/users/index.html.haml [hunk 7]
```diff
@@ -1,2 +1,36 @@
...
+              #{user.first_name}
```
Comment (igorkasyanchuk) [inline app/views/account/users/index.html.haml]: output full_name

File: spec/features/invitation_feature_spec.rb [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        visit_in_email 'Accept invitation'
+
+        fill_in 'First Name', with: 'name'
+        fill_in 'Last Name', with: 'last_name'
+        fill_in 'Password', with: 'password'
+        fill_in 'Confirm Password', with: 'password'
+
+        click_button 'Continue'
+
+        sign_in user
```
Comment (igorkasyanchuk) [inline spec/features/invitation_feature_spec.rb]: why you have this sign_in user here?

Where user is located after continue? on which page?

File: app/assets/stylesheets/spacex/custom_styles.scss [hunk 9]
```diff
@@ -6,3 +6,9 @@
 .invalid-feedback {
   color: red;
 }
+
+#invitation-form {
+  .form-group input{
+      margin-left: 10px;
```
Comment (igorkasyanchuk) [inline app/assets/stylesheets/spacex/custom_styles.scss]: formatting

File: spec/features/invitation_feature_spec.rb (hunk 8 above)
Comment (slvk14) [reply spec/features/invitation_feature_spec.rb]: User that is signed in after 'Continue'  is on the root_path, but it is a member and it doesn't have access to the account_users_path, so i sign in again as admin that invited this user to check status

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -6,4 +6,10 @@ def configure_permitted_parameters
     devise_parameter_sanitizer.permit(:invite, keys: %i[email organization_id])
     devise_parameter_sanitizer.permit(:sign_up, keys: [:first_name, :last_name, :terms_of_service, own_organization_attributes: [:name]])
   end
+
+  private
+
+  def require_admin
+    redirect_to root_path unless current_user.role == 'admin'
```
Comment (romaslmd) [inline app/controllers/application_controller.rb]: change role check to `current_user.admin?`

Comment (romaslmd) [reply app/controllers/application_controller.rb]: + add some flash message

File: app/models/user.rb [hunk 11]
```diff
@@ -15,12 +15,20 @@ class User < ApplicationRecord
...
-  after_create :send_signup_emails
+  after_create :send_signup_emails, :set_admin_role
 
   def full_name
     [first_name, last_name].reject(&:blank?).join(' ')
   end
 
+  def status
+    if self.invitation_accepted?
+      return "accepted"
```
Comment (romaslmd) [inline app/models/user.rb]: redundant `return` statement

File: app/models/user.rb [hunk 12]
```diff
@@ -30,4 +38,12 @@ def send_signup_emails
       SignupMailer.new_organization_created_email(self, super_admin).deliver_now
     end
   end
+
+  def set_admin_role
+    if self.own_organization.present?
+      self.role = 'admin'
+      self.save
```
Comment (romaslmd) [inline app/models/user.rb]: does this require validations and callbacks?

File: app/views/account/users/index.html.haml [hunk 13]
```diff
@@ -13,24 +13,19 @@
       - @users.each do |user|
         %tr
           %td 
-            - if user.role == 'admin'
-              #{user.first_name} 
+            - if user.admin?
+              #{user.full_name} 
               %strong (admin)
             - elsif user.invitation_accepted? == false
```
Comment (romaslmd) [inline app/views/account/users/index.html.haml]: what's this doing here?

File: app/models/user.rb [hunk 14]
```diff
@@ -30,4 +38,12 @@ def send_signup_emails
       SignupMailer.new_organization_created_email(self, super_admin).deliver_now
     end
   end
+
+  def set_admin_role
+    if self.own_organization.present?
```
Comment (romaslmd) [inline app/models/user.rb]: prefer 
```
return if own_organization.ablank?
update_attribute :role, 'admin'
self.own_organization.users.push(self)
```

File: app/controllers/application_controller.rb (hunk 10 above)
Comment (igorkasyanchuk) [reply app/controllers/application_controller.rb]: agree with Roman

File: app/controllers/application_controller.rb [hunk 15]
```diff
@@ -10,6 +10,9 @@ def configure_permitted_parameters
   private
 
   def require_admin
-    redirect_to root_path unless current_user.role == 'admin'
+    unless current_user.admin?
+    redirect_to root_path 
+    flash.now[:error] = 'Sorry, you have to be admin to continue..'
```
Comment (romaslmd) [inline app/controllers/application_controller.rb]: care about the order

File: app/controllers/users/invitations_controller.rb [hunk 16]
```diff
@@ -0,0 +1,113 @@
+class Users::InvitationsController < Devise::InvitationsController
+  prepend_before_action :authenticate_inviter!, only: %i[new create]
+  prepend_before_action :has_invitations_left?, only: [:create]
+  prepend_before_action :require_no_authentication, only: %i[edit update destroy]
+  prepend_before_action :resource_from_invitation_token, only: %i[edit destroy]
+
+  helper_method :after_sign_in_path_for if respond_to? :helper_method
```
Comment (romaslmd) [inline app/controllers/users/invitations_controller.rb]: what's w/ `if` condition?

Comment (romaslmd) [reply app/controllers/users/invitations_controller.rb]: why do u override the implementation?

File: app/controllers/users/invitations_controller.rb [hunk 17]
```diff
@@ -0,0 +1,113 @@
+class Users::InvitationsController < Devise::InvitationsController
+  prepend_before_action :authenticate_inviter!, only: %i[new create]
```
Comment (romaslmd) [inline app/controllers/users/invitations_controller.rb]: don't override gem implementation

File: app/controllers/users/invitations_controller.rb [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+  def update
+    raw_invitation_token = update_resource_params[:invitation_token]
+    self.resource = accept_resource
+    invitation_accepted = resource.errors.empty?
+
+    yield resource if block_given?
+
+    if invitation_accepted
+      if resource.class.allow_insecure_sign_in_after_accept
+        flash_message = resource.active_for_authentication? ? :updated : :updated_not_active
```
Comment (romaslmd) [inline app/controllers/users/invitations_controller.rb]: remove all this, override only those u need to, in other case - call `super` or simply inherit the implementation class
never override the implementation of the gem unless it's necessary!

File: app/views/shared/dashboard/_menu.html.haml [hunk 19]
```diff
@@ -28,4 +28,9 @@
           %li
             = link_to account_device_groups_path do
               %i.menu-icon.ti-desktop
-              %span Device Groups
\ No newline at end of file
+              %span Device Groups
+          - if current_user.role == 'admin'
```
Comment (romaslmd) [inline app/views/shared/dashboard/_menu.html.haml]: `admin?`

File: db/migrate/20190606114651_remove_organization_id_column_from_user.rb [hunk 20]
```diff
@@ -0,0 +1,5 @@
+class RemoveOrganizationIdColumnFromUser < ActiveRecord::Migration[5.2]
+  def change
+    remove_column :users, :organization_id, :integer
```
Comment (romaslmd) [inline db/migrate/20190606114651_remove_organization_id_column_from_user.rb]: why do u need 2 separate migration to perform 1 action?
removing the column automatically removes all related indexes

File: db/migrate/20190606115037_add_organization_id_to_user.rb [hunk 21]
```diff
@@ -0,0 +1,5 @@
+class AddOrganizationIdToUser < ActiveRecord::Migration[5.2]
+  def change
+    add_reference :users, :organization
```
Comment (romaslmd) [inline db/migrate/20190606115037_add_organization_id_to_user.rb]: are u aware that removing the organization_id column will delete all data?

Comment (slvk14) [reply db/migrate/20190606115037_add_organization_id_to_user.rb]: So should i change smth?

File: app/controllers/application_controller.rb [hunk 22]
```diff
@@ -10,6 +10,9 @@ def configure_permitted_parameters
   private
 
   def require_admin
-    redirect_to root_path unless current_user.role == 'admin'
+    unless current_user.admin?
+    redirect_to root_path
```
Comment (igorkasyanchuk) [inline app/controllers/application_controller.rb]: code formatting + what Roman said

File: db/migrate/20190606194808_change_default_user_role_back.rb [hunk 23]
```diff
@@ -0,0 +1,5 @@
+class ChangeDefaultUserRoleBack < ActiveRecord::Migration[5.2]
+  def change
+    change_column_default(:users, :role, from: 1, to: 0)
```
Comment (romaslmd) [inline db/migrate/20190606194808_change_default_user_role_back.rb]: why r u changing default of the same column in 2 different migrations in the same PR?

File: db/migrate/20190606115037_add_organization_id_to_user.rb (hunk 21 above)
Comment (romaslmd) [reply db/migrate/20190606115037_add_organization_id_to_user.rb]: depends if u want to keep existing data or not

File: db/migrate/20190606194808_change_default_user_role_back.rb (hunk 23 above)
Comment (slvk14) [reply db/migrate/20190606194808_change_default_user_role_back.rb]: So can i simply delete them? I just thought its not a good practice to delete existing migrations

Comment (romaslmd) [reply db/migrate/20190606194808_change_default_user_role_back.rb]: what do u mean by "existing"? as i see the `db/migrate/20190605150126_change_default_user_role.rb` is added in this PR

**Pull request merged**